Migration OCP 3 to OCP 4 (#108)
mramendi/kie-docs: doc-content/enterprise-only/openshift/migration-ocp3-4-mysql-proc.adoc

.Prerequisites

* You must have a machine that has network access to both the {OPENSHIFT} version 3 and {OPENSHIFT} version 4 infrastructures.
* The `oc` command-line client from {OPENSHIFT} version 4 must be installed on the machine.
* The `mysql` and `mysqldump` client applications provided by MySQL version 8 or later or by MariaDB version 10 or later must be installed.

.Procedure

. Ensure that no web clients and no client applications are connected to any elements of the old and new deployment, including {CENTRAL} and {KIE_SERVERS}.
. Create a temporary directory and change into this directory.
. Using the `oc` command, log in to the {OPENSHIFT} version 3 infrastructure and switch to the project containing the old deployment.
. To view the deployment configuration names in the old deployment, run the command:
+
----
oc get dc
----
+
Find the `mysql` deployment configuration that corresponds to the {KIE_SERVER}.
+
. View the configuration YAML of the deployment configuration, for example:
+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----

> oc edit dc/myapp-mysql

Reviewer: If you want a oneliner for getting the values:
```
oc get dc/myapp-mysql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="MYSQL_USER")]}'.value
oc get dc/myapp-mysql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="MYSQL_PASSWORD")]}'.value
```

[id='migration-ocp3-4-mysql-proc-{context}']
= Migrating a MySQL database for a {KIE_SERVER}

If your environment in {OPENSHIFT} version 3 includes a {KIE_SERVER} that uses a MySQL database pod, copy the MySQL database content from the old deployment to the new deployment. This action copies the existing process state to the new deployment.

.Prerequisites

* You must have a machine that has network access to both the {OPENSHIFT} version 3 and {OPENSHIFT} version 4 infrastructures.
* The `oc` command-line client from {OPENSHIFT} version 4 must be installed on the machine. For instructions about installing the command-line client, see https://access.redhat.com/documentation/en-us/openshift_container_platform/4.4/html/cli_tools/openshift-cli-oc#installing-the-cli[_CLI tools_] in {OPENSHIFT} documentation.
* The `mysql` and `mysqldump` client applications provided by MySQL version 8 or later or by MariaDB version 10 or later must be installed.

.Procedure

. Ensure that no web clients and no client applications are connected to any elements of the old and new deployment, including {CENTRAL} and {KIE_SERVERS}.

> . Create a temporary directory and change into this directory.

Reviewer 2: Change into this directory? this sounds confusing. Are we changing to the new directory?

. Using the `oc` command, log in to the {OPENSHIFT} version 3 infrastructure and switch to the project containing the old deployment.

> . To view the deployment configuration names in the old deployment, run the command:

Reviewer 2: ```suggestion
. To view the deployment configuration names in the old deployment, run the following command:
```

+
----
oc get dc
----
+
Find the `mysql` deployment configuration that corresponds to the {KIE_SERVER}.
+
. View the configuration YAML of the deployment configuration, for example:
+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----
oc edit dc/myapp-mysql
----
+
In this file, find the user name (normally `{PRODUCT_INIT}`) and password for the database server, for example:
+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----
- name: MYSQL_USER
  value: {PRODUCT_INIT}
- name: MYSQL_PASSWORD
  value: NDaJIV7!
----
+
Record the user name and password. Do not make any changes to the file.
+
[NOTE]
====
You can also use the following commands to retrieve the user name and password:
----
oc get dc/myapp-postgresql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="POSTGRESQL_USER")]}'.value

oc get dc/myapp-postgresql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="POSTGRESQL_PASSWORD")]}'.value
----
====
+

> . To view the service names in the old deployment, run the command:

Reviewer 2: ```suggestion
. To view the service names in the old deployment, run the following command:
```

+
----
oc get svc
----
+
Find the `mysql` service that corresponds to the {KIE_SERVER}.
+
. In a separate terminal window, start port forwarding from the local host to the `mysql` service, using the name and port number displayed for the service, for example:
+
----
oc port-forward service/myapp-mysql 3306:3306
----
+

> . Create a full database dump using the recorded username, for example:

Reviewer 2: ```suggestion
. Create a full database dump using the recorded user name, for example:
```

+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----
mysqldump --all-databases -u {PRODUCT_INIT} -p -h 127.0.0.1 > mysqldump.sql
----
+
When prompted, enter the recorded password. The dump creation can take considerable time.
+ 

> . Stop the port forwarding in the separate window using the `Ctrl`+`C` key combination.

Reviewer 2: I'd suggest using keyboard shortcut syntax. 
```suggestion
. Stop the port forwarding in the separate window using the kbd:[Ctrl+C] key combination.
```

. Using the `oc` command, log in to the {OPENSHIFT} version 4 infrastructure and switch to the project containing the new deployment.

> . To view the deployment configuration names in the new deployment, run the command:

Reviewer 2: ```suggestion
. To view the deployment configuration names in the new deployment, run the following command:
```

+
----
oc get dc
----
+
Find the `mysql` deployment configuration that corresponds to the {KIE_SERVER}.
+
. View the configuration YAML of the deployment configuration, for example:
+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----
oc edit dc/myappnew-mysql
----
+
In this file, find the user name (normally `{PRODUCT_INIT}`) and password for the database server. Record the user name and password. Do not make any changes to the file.
+
[NOTE]
====
You can also use the following commands to retrieve the user name and password:
----
oc get dc/myapp-postgresql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="POSTGRESQL_USER")]}'.value

oc get dc/myapp-postgresql -ojsonpath='{.spec.template.spec.containers[0].env[?(@.name=="POSTGRESQL_PASSWORD")]}'.value
----
====
+

> . To view the service names in the new deployment, run the command:

Reviewer 2: ```suggestion
. To view the service names in the new deployment, run the following command:
```

+
----
oc get svc
----
+
Find the `mysql` service that corresponds to the {KIE_SERVER}.
+
. In a separate terminal window, start port forwarding from the local host to the `mysql` service, using the name and port number displayed for the service, for example:
+
----
oc port-forward service/myappnew-mysql 3306:3306
----
+

> . Restore the database dump using the recorded username, for example:

Reviewer 2: ```suggestion
. Restore the database dump using the recorded user name, for example:
```

+
[subs="attributes,verbatim,macros,specialchars,replacements"]
----
mysql -u {PRODUCT_INIT} -p -h 127.0.0.1 < mysqldump.sql
----
+
When prompted, enter the recorded password. The restoration can take considerable time.
+ 

> . Stop the port forwarding in the separate window using the `Ctrl`+`C` key combination.

Reviewer 2: ```suggestion
. Stop the port forwarding in the separate window using the kbd:[Ctrl+C] key combination.
```
